syntax = "proto3";

package test;

import "google/protobuf/duration.proto";
import "include/stout/flags/v1/flag.proto";
import "include/stout/flags/v1/positional_argument.proto";
import "include/stout/flags/v1/subcommand.proto";

///////////////////////////////////////////////////////////////////////////////

message Flags {
  string foo = 1 [
    (stout.v1.flag) = {
      names: [ "foo" ]
      help: "help"
      required: true
    }
  ];

  string _s = 2 [
    (stout.v1.flag) = {
      names: [ "s" ]
      help: "help"
      required: false
    }
  ];

  bool bar = 3 [
    (stout.v1.flag) = {
      names: [ "bar" ]
      help: "help"
    }
  ];

  int32 baz = 4 [
    (stout.v1.flag) = {
      names: [ "baz" ]
      help: "help"
    }
  ];

  google.protobuf.Duration duration = 5 [
    (stout.v1.flag) = {
      names: [ "duration" ]
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message MissingFlagName {
  string s = 1 [
    (stout.v1.flag) = {
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message MissingFlagHelp {
  string s = 1 [
    (stout.v1.flag) = {
      names: "name"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message DuplicateFlagName {
  string s1 = 1 [
    (stout.v1.flag) = {
      names: [ "same" ]
      help: "help"
    }
  ];

  string s2 = 2 [
    (stout.v1.flag) = {
      names: [ "other", "same" ]
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message DuplicateFlagsDeath {
  string s = 1 [
    (stout.v1.flag) = {
      names: [ "s" ]
      help: "help"
    }
  ];

  string ss = 2 [
    (stout.v1.flag) = {
      names: [ "ss", "ss_alias" ]
      help: "help"
    }
  ];

  bool duplicate = 3 [
    (stout.v1.flag) = {
      names: [ "duplicate", "other" ]
      help: "help"
    }
  ];

  bool b = 4 [
    (stout.v1.flag) = {
      names: [ "b" ]
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message BuildFlag {
  int32 other_flag = 1 [
    (stout.v1.flag) = {
      names: [ "other_flag" ]
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message InfoFlag {
  string info = 1 [
    (stout.v1.flag) = {
      names: [ "info" ]
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message FlagsWithSubcommandMissingName {
  oneof subcommand {
    BuildFlag build = 1 [
      (stout.v1.subcommand) = {
        help: "help"
      }
    ];
    InfoFlag info_subcommand = 2 [
      (stout.v1.subcommand) = {
        names: [ "info_subcommand" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message FlagsWithSubcommandMissingHelp {
  oneof subcommand {
    BuildFlag build = 1 [
      (stout.v1.subcommand) = {
        names: [ "build" ]
        help: "help"
      }
    ];
    InfoFlag info_subcommand = 2 [
      (stout.v1.subcommand) = {
        names: [ "info_subcommand" ]
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message SubcommandFlagsWithoutExtension {
  bool b = 1 [
    (stout.v1.flag) = {
      names: [ "b" ]
      help: "help"
    }
  ];

  oneof subcommand {
    BuildFlag build = 2 [
      (stout.v1.subcommand) = {
        names: [ "build" ]
        help: "help"
      }
    ];
    InfoFlag info_subcommand = 3;
  }
}

///////////////////////////////////////////////////////////////////////////////

message FlagsWithIncorrectExtension {
  bool b = 1 [
    (stout.v1.subcommand) = {
      names: [ "build" ]
      help: "help"
    }
  ];

  oneof subcommand {
    BuildFlag build = 2 [
      (stout.v1.subcommand) = {
        names: [ "build" ]
        help: "help"
      }
    ];
    InfoFlag info_subcommand = 3 [
      (stout.v1.subcommand) = {
        names: [ "info_subcommand" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message IncorrectOneofName {
  oneof other {
    BuildFlag build = 1 [
      (stout.v1.subcommand) = {
        names: [ "build" ]
        help: "help"
      }
    ];
    InfoFlag info_subcommand = 2 [
      (stout.v1.subcommand) = {
        names: [ "info_subcommand" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message SubcommandFlagExtension {
  oneof subcommand {
    BuildFlag build = 1 [
      (stout.v1.flag) = {
        names: [ "build" ]
        help: "help"
      }
    ];
    InfoFlag info_subcommand = 2 [
      (stout.v1.subcommand) = {
        names: [ "info_subcommand" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message SimpleSubcommandSucceed {
  bool b = 1 [
    (stout.v1.flag) = {
      names: [ "b", "bb" ]
      help: "help"
    }
  ];

  oneof subcommand {
    BuildFlag build = 2 [
      (stout.v1.subcommand) = {
        names: [ "build" ]
        help: "help"
      }
    ];
    InfoFlag info_subcommand = 3 [
      (stout.v1.subcommand) = {
        names: [ "info_subcommand" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message ComplicatedSubcommandMessage {
  string flag = 1 [
    (stout.v1.flag) = {
      names: [ "flag" ]
      help: "help"
    }
  ];

  string other = 2 [
    (stout.v1.flag) = {
      names: [ "other" ]
      help: "help"
    }
  ];

  oneof subcommand {
    SubcommandSubMessage1 sub1 = 3 [
      (stout.v1.subcommand) = {
        names: [ "sub1" ]
        help: "help"
      }
    ];
    SubcommandSubMessage2 sub2 = 4 [
      (stout.v1.subcommand) = {
        names: [ "sub2" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message SubcommandSubMessage1 {
  string another = 1 [
    (stout.v1.flag) = {
      names: [ "another" ]
      help: "help"
    }
  ];

  int32 num = 2 [
    (stout.v1.flag) = {
      names: [ "num" ]
      help: "help"
    }
  ];

  oneof subcommand {
    BuildFlag build = 3 [
      (stout.v1.subcommand) = {
        names: [ "build" ]
        help: "help"
      }
    ];
    InfoFlag info_subcommand = 4 [
      (stout.v1.subcommand) = {
        names: [ "info_subcommand" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message SubcommandSubMessage2 {
  string s = 1 [
    (stout.v1.flag) = {
      names: [ "s" ]
      help: "help"
    }
  ];

  oneof subcommand {
    BuildFlag build = 2 [
      (stout.v1.subcommand) = {
        names: [ "build" ]
        help: "help"
      }
    ];
    InfoFlag info_subcommand = 3 [
      (stout.v1.subcommand) = {
        names: [ "info_subcommand" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message DuplicateEnclosingFlagName {
  int32 other_flag = 1 [
    (stout.v1.flag) = {
      names: [ "other_flag" ]
      help: "help"
    }
  ];

  oneof subcommand {
    BuildFlag build = 2 [
      (stout.v1.subcommand) = {
        names: [ "build" ]
        help: "help"
      }
    ];
    InfoFlag info_subcommand = 3 [
      (stout.v1.subcommand) = {
        names: [ "info_subcommand" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message DuplicateSubcommandFields {
  oneof subcommand {
    BuildFlag build = 1 [
      (stout.v1.subcommand) = {
        names: [ "build" ]
        help: "help"
      }
    ];
    InfoFlag info = 2 [
      (stout.v1.subcommand) = {
        names: [ "build" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message Rename {
  string cur_file_name = 1 [
    (stout.v1.argument) = {
      name: "current_file"
      help: "help"
    }
  ];

  string new_file_name = 2 [
    (stout.v1.argument) = {
      name: "new_file"
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message BuildFile {
  bool debug_mode = 1 [
    (stout.v1.flag) = {
      names: [ "debug" ]
      help: "help"
    }
  ];

  string file = 2 [
    (stout.v1.argument) = {
      name: "file_name"
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message ProcessFile {
  oneof subcommand {
    BuildFile build = 1 [
      (stout.v1.subcommand) = {
        names: [ "build" ]
        help: "help"
      }
    ];

    Rename rename = 2 [
      (stout.v1.subcommand) = {
        names: [ "rename" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

// This message is illegal because we expect that users
// will define fields with 'stout.v1.argument' extension
// of the type 'string'.
message IllegalPositionalArg1 {
  int32 num = 1 [
    (stout.v1.argument) = {
      name: "num"
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

// Illegal setup where the user doesn't define name option.
message IllegalPositionalArg2 {
  string str = 1 [
    (stout.v1.argument) = {
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

// Illegal setup where the user doesn't define help option.
message IllegalPositionalArg3 {
  string str = 1 [
    (stout.v1.argument) = {
      name: "str"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

// This message demonstrates that if users define positional
// arguments with random field index order - anyway the order
// of positional arguments that we expect from the command
// line will be the same as the order in which they are
// defined in protobuf message.
message RandomOrderFieldIndexes {
  string str1 = 2 [
    (stout.v1.argument) = {
      name: "str1"
      help: "help"
    }
  ];

  string str2 = 3 [
    (stout.v1.argument) = {
      name: "str2"
      help: "help"
    }
  ];

  string str3 = 1 [
    (stout.v1.argument) = {
      name: "str3"
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

// More complicated case for positional arguments.
message TopLevelArguments {
  int32 top_level_flag1 = 1 [
    (stout.v1.flag) = {
      names: [ "top_level_flag1" ]
      help: "help"
    }
  ];

  string top_level_flag2 = 2 [
    (stout.v1.flag) = {
      names: [ "top_level_flag2" ]
      help: "help"
    }
  ];

  string top_pos_arg1 = 3 [
    (stout.v1.argument) = {
      name: "top_pos_arg1"
      help: "help"
    }
  ];

  string top_pos_arg2 = 4 [
    (stout.v1.argument) = {
      name: "top_pos_arg2"
      help: "help"
    }
  ];

  oneof subcommand {
    Subcommand1 sub1 = 5 [
      (stout.v1.subcommand) = {
        names: [ "sub1" ]
        help: "help"
      }
    ];
    Subcommand2 sub2 = 6 [
      (stout.v1.subcommand) = {
        names: [ "sub2" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message Subcommand1 {
  int32 sub1_flag = 1 [
    (stout.v1.flag) = {
      names: [ "sub1_flag" ]
      help: "help"
    }
  ];

  string sub1_pos_arg = 2 [
    (stout.v1.argument) = {
      name: "sub1_pos_arg1"
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message Subcommand2 {
  int32 sub2_flag = 1 [
    (stout.v1.flag) = {
      names: [ "sub2_flag" ]
      help: "help"
    }
  ];

  string sub2_pos_arg = 2 [
    (stout.v1.argument) = {
      name: "sub2_pos_arg"
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////
